cli/standalone/serialization.rs

```diff
@@ -42,50 +43,46 @@ pub fn serialize_binary_data_section(
   remote_modules: &RemoteModulesStoreBuilder,
   vfs: &BuiltVfs,
 ) -> Result<Vec<u8>, AnyError> {
-  fn write_bytes_with_len(bytes: &mut Vec<u8>, data: &[u8]) {
-    bytes.extend_from_slice(&(data.len() as u64).to_le_bytes());
-    bytes.extend_from_slice(data);
-  }
-
-  let mut bytes = Vec::new();
-  bytes.extend_from_slice(MAGIC_BYTES);
-
-  // 1. Metadata
-  {
-    let metadata = serde_json::to_string(metadata)?;
-    write_bytes_with_len(&mut bytes, metadata.as_bytes());
-  }
-  // 2. Npm snapshot
-  {
-    let npm_snapshot =
-      npm_snapshot.map(serialize_npm_snapshot).unwrap_or_default();
-    write_bytes_with_len(&mut bytes, &npm_snapshot);
-  }
-  // 3. Remote modules
-  {
-    let update_index = bytes.len();
-    bytes.extend_from_slice(&(0_u64).to_le_bytes());
-    let start_index = bytes.len();
-    remote_modules.write(&mut bytes)?;
-    let length = bytes.len() - start_index;
-    let length_bytes = (length as u64).to_le_bytes();
-    bytes[update_index..update_index + length_bytes.len()]
-      .copy_from_slice(&length_bytes);
-  }
-  // 4. VFS
-  {
-    let serialized_vfs = serde_json::to_string(&vfs.root)?;
-    write_bytes_with_len(&mut bytes, serialized_vfs.as_bytes());
-    let vfs_bytes_len = vfs.files.iter().map(|f| f.len() as u64).sum::<u64>();
-    bytes.extend_from_slice(&vfs_bytes_len.to_le_bytes());
-    for file in &vfs.files {
-      bytes.extend_from_slice(file);
+  let metadata = serde_json::to_string(metadata)?;
+  let npm_snapshot =
+    npm_snapshot.map(serialize_npm_snapshot).unwrap_or_default();
+  let remote_modules_len = Cell::new(0_u64);
+  let serialized_vfs = serde_json::to_string(&vfs.root)?;
+
+  let bytes = capacity_builder::BytesBuilder::build(|builder| {
+    builder.append(MAGIC_BYTES);
+    // 1. Metadata
+    {
+      builder.append_le(metadata.len() as u64);
+      builder.append(&metadata);
+    }
+    // 2. Npm snapshot
+    {
+      builder.append_le(npm_snapshot.len() as u64);
+      builder.append(&npm_snapshot);
+    }
+    // 3. Remote modules
+    {
+      builder.append_le(remote_modules_len.get()); // this will be properly initialized on the second pass
+      let start_index = builder.len();
+      remote_modules.write(builder);
+      remote_modules_len.set((builder.len() - start_index) as u64);
+    }
+    // 4. VFS
+    {
+      builder.append_le(serialized_vfs.len() as u64);
+      builder.append(&serialized_vfs);
+      let vfs_bytes_len = vfs.files.iter().map(|f| f.len() as u64).sum::<u64>();
+      builder.append_le(vfs_bytes_len);
+      for file in &vfs.files {
+        builder.append(file);
+      }
     }
-  }
 
-  // write the magic bytes at the end so we can use it
-  // to make sure we've deserialized correctly
-  bytes.extend_from_slice(MAGIC_BYTES);
+    // write the magic bytes at the end so we can use it
+    // to make sure we've deserialized correctly
+    builder.append(MAGIC_BYTES);
+  })?;
```

Comment:
We also now surface if the capacity couldn't be allocated as an error instead of panicking.